spec/integration/application/agent_spec.rb

```diff
@@ -0,0 +1,257 @@
+require 'spec_helper'
+require 'puppet_spec/files'
+require 'puppet_spec/puppetserver'
+require 'puppet_spec/compiler'
+
+describe "puppet agent", unless: Puppet::Util::Platform.jruby? do
+  include PuppetSpec::Files
+
+  before :all do
+    WebMock.disable!
+  end
+
+  after :all do
+    WebMock.enable!
+  end
+
+  before :each do
+    # make sure we don't take too long
+    Puppet[:http_connect_timeout] = '5s'
+    Puppet[:server] = '127.0.0.1'
+    Puppet[:certname] = '127.0.0.1'
+
+    Puppet[:localcacert] = File.join(PuppetSpec::FIXTURE_DIR, 'ssl', 'ca.pem')
+    Puppet[:hostcrl] = File.join(PuppetSpec::FIXTURE_DIR, 'ssl', 'crl.pem')
+    Puppet[:hostprivkey] = File.join(PuppetSpec::FIXTURE_DIR, 'ssl', '127.0.0.1-key.pem')
+    Puppet[:hostcert] = File.join(PuppetSpec::FIXTURE_DIR, 'ssl', '127.0.0.1.pem')
+
+    facts = Puppet::Node::Facts.new(Puppet[:certname])
+    Puppet::Node::Facts.indirection.save(facts)
+  end
+
+  let(:server) { PuppetSpec::Puppetserver.new }
+  let(:agent) { Puppet::Application[:agent] }
+  let(:logfile) { tmpfile('agent.log') }
+  let(:args) { ['--test', '--logdest', logfile] }
+
+  context 'server_list' do
+    it "uses the first server in the list" do
+      Puppet[:server_list] = '127.0.0.1'
+
+      server.start_server do |port|
+        Puppet[:masterport] = port
+        expect {
+          agent.command_line.args = args
+          agent.run
+        }.to exit_with(0)
+      end
+    end
+
+    it "falls back, recording the first viable server in the report" do
+      server.start_server do |port|
+        Puppet[:server_list] = "puppet.example.com,#{Puppet[:server]}"
+        Puppet[:masterport] = port
+        expect {
+          agent.command_line.args = args
+          agent.run
+        }.to exit_with(0)
+
+        report = Puppet::Transaction::Report.convert_from(:yaml, File.read(Puppet[:lastrunreport]))
+        expect(report.master_used).to eq("127.0.0.1:#{port}")
+      end
+    end
+
+    it "omits master_used it no servers could be contacted" do
+      # PENDING: we should still write a report and try to send it
+      Puppet[:server_list] = "puppet.example.com"
+
+      # This doesn't actually need a server, because we never send the report
+      # but we should send one.
```

Comment:
Does 5.5.x behave the same way -- not sending a report if the server list is exhausted?

Reply:
Yep it's the same